src/auth/discord.rs

```diff
@@ -1 +1,404 @@
+    pub grant_type: String,
+    pub code: String,
+    pub redirect_uri: String,
+}
+pub(crate) enum DiscordScopes {
+    Identify,
+    Guilds,
+    GuildsMembersRead,
+}
 
+impl DiscordScopes {
+    pub(crate) fn as_str(&self) -> &'static str {
+        match self {
+            DiscordScopes::Identify => "identify",
+            DiscordScopes::Guilds => "guilds",
+            DiscordScopes::GuildsMembersRead => "guilds.members.read",
+        }
+    }
+}
+// route: /auth/login/
+pub async fn login(req: actix_web::HttpRequest) -> fpm::Result<fpm::http::Response> {
+    // Discord will be redirect to this url after login process completed
+
+    let redirect_url: String = format!(
+        "{}://{}{}",
+        req.connection_info().scheme(),
+        req.connection_info().host(),
+        CALLBACK_URL
+    );
+    let client_id = match std::env::var("DISCORD_CLIENT_ID") {
+        Ok(id) => id,
+        Err(_e) => {
+            println!("WARN: DISCORD_CLIENT_ID not set");
+            // TODO: Need to change this approach later
+            "FPM_TEMP_DISCORD_CLIENT_ID".to_string()
+        }
+    };
+    let discord_auth_url = format!(
+        "{}{}{}{}{}{}{}{}{}{}{}",
+        AUTH_URL,
+        "?client_id=",
+        client_id,
+        "&redirect_uri=",
+        redirect_url,
+        "&response_type=code&scope=",
+        DiscordScopes::Identify.as_str(),
+        " ",
+        DiscordScopes::Guilds.as_str(),
+        " ",
+        DiscordScopes::GuildsMembersRead.as_str()
+    );
+    // send redirect to /auth/discord/callback/
+    Ok(actix_web::HttpResponse::Found()
+        .append_header((actix_web::http::header::LOCATION, discord_auth_url))
+        .finish())
+}
+
+// route: /auth/discord/callback/
+// In this API we are accessing
+// the token and setting it to cookies
+pub async fn callback(req: actix_web::HttpRequest) -> fpm::Result<actix_web::HttpResponse> {
+    use magic_crypt::MagicCryptTrait;
+    #[derive(Debug, serde::Deserialize)]
+    pub struct QueryParams {
+        pub code: String,
+    }
+
+    let query = actix_web::web::Query::<QueryParams>::from_query(req.query_string())?.0;
+    let redirect_url = format!(
+        "{}://{}{}",
+        req.connection_info().scheme(),
+        req.connection_info().host(),
+        CALLBACK_URL
+    );
+    let discord_auth =
+        apis::discord_token(TOKEN_URL, redirect_url.as_str(), query.code.as_str()).await;
+    match discord_auth {
+        Ok(access_token) => {
+            dbg!(&access_token);
+            let (user_name, user_id) = apis::user_details(&access_token).await?;
+            let user_detail_obj: UserDetail = UserDetail {
+                token: access_token.to_owned(),
+                user_name,
+                user_id,
+            };
+            let user_detail_str = serde_json::to_string(&user_detail_obj)?;
+            let secret_key = fpm::auth::secret_key();
+            let mc_obj = magic_crypt::new_magic_crypt!(secret_key.as_str(), 256);
```

Comment:
Move `magic_crypt` logic to utility, as this is going to use in each auth provider and the logic is going to be the same for each auth provider.

Reply:
done